README.md
  
[![Build Status](https://travis-ci.org/mjirik/exsu.svg?branch=master)](https://travis-ci.org/mjirik/exsu)
[![Coverage Status](https://coveralls.io/repos/github/mjirik/exsu/badge.svg?branch=master)](https://coveralls.io/github/mjirik/exsu?branch=master)
[![PyPI version](https://badge.fury.io/py/exsu.svg)](http://badge.fury.io/py/exsu)


# exsu

Experiment support tools prepared for computer vision experiments. 
Output directory and spreadsheet file is managed by package.

* Spreadsheet data and report images are stored into output directory
* Images are reported according a selected severity
* Common spreadseet containing complete history from all output directories can be produced
* Version of particular python package can be reported in the spreadsheet
* Repository status, git hash and dirty files can be reported in the spreadsheet

## Write spreadsheet data

```python
from pathlib import Path
import exsu

outputdir = Path("./test_report/")
commonsheet = Path("./test_report_common_spreadsheet.xlsx")

report = exsu.report.Report(outputdir=outputdir, additional_spreadsheet_fn=commonsheet)
report.add_cols_to_actual_row({"Col1": 25, "Col2": "test string", "Col5": 5})
report.add_cols_to_actual_row({"Col2": "prepsanu", "Col1": 26, "Col4": "ctyrka"})
report.finish_actual_row()

report.add_cols_to_actual_row({"Col1": 27, "Col2": "test string", "Col3": "trojka"})
report.finish_actual_row()

# Save all into `outputdir` and also into `commonsheet`
report.dump()

# new write to common excel
report.init()
report.add_cols_to_actual_row({"Col1": 28, "Col2": "new line to common", "Col7": 77})
report.finish_actual_row()
report.dump()

```

## Write image data

Write image data into output directory

```python
from pathlib import Path
import numpy as np
import exsu

outputdir = Path("./test_report/")
commonsheet = Path("./test_report_common_spreadsheet.xlsx")

fn = "test_image.png"
img = 50 + np.random.rand(100, 100) * 30
img[20:60, 20:60] += 100
img = img.astype(np.uint8)
report = exsu.report.Report(outputdir=outputdir, additional_spreadsheet_fn=commonsheet)
report.imsave(fn, img)
```


## Git repository info

```python
from pathlib import Path
import numpy as np
import exsu
report = exsu.report.Report(repodir=Path(__file__).parent.resolve())
report.finish_actual_row()
report.df
```
Output DataFrame
```
                               repo exsu id  ...  repo exsu dirty files
0  1da301da36cec0fb931f60dd9acca790ec715892  ...      exsu/git_tools.py
```

More than one repositories can be added with `report.add_repo(repodir, reponame=None)`.


## Automatically report versions of selected packages

The version of crucial packages can be reported with the data.

```python
import exsu
report = exsu.report.Report(check_version_of=["numpy", "scipy"])
report.finish_actual_row()
report.df

```


changelog.md
# Changelog

Actual version 0.7.0

## 0.6

 * [Added] `savefig_no_border` added to save figure with no borders at all.
 * [Changed] Additional `kwargs` passed through `savefig` functions.
 * [Added] Create file with name `{stem}.1.suffix` if spreadsheet file is opened in different application

## 0.5
 * [Added] CSV support
 * [Changed] Additional spreadsheet saved without index to be compatible with common spreadsheet
 * [Added] Find if of closest points in the dataframe
 * [Added] `kwargs` can be used in `imsave()`
 * [Added] Flatten keys in struct. The list and ndarray can be converted to simple cols.

## Version 0.4.

 * [Added] Check version of selected packages
 * [Added] Added functions for work with parameters
 * [Added] Added icon cration function
 * [Changed] Outputdir is created in the moment it is required. 

## Version 0.3.

 * [Added] More function to save image and figure. 

## Version 0.2.

 * [Added] Git repository version and describe added to report

## Version 0.1. 

 * [Added] Report support
 * [Added] Clean persistent cols ability


tests/dili_test.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8

from loguru import logger
import numpy as np

import unittest
from collections import OrderedDict
from exsu import dili
import pytest


def test_get_fcn_params():
    def test_fcn(a, b, c=1, d=2):
        return a + b + c

    class MyTestObj:
        def __init__(self, oa, ob, oc=1, od=2):
            pass

    myobj = MyTestObj(1, 2)
    argsf = dili.get_default_kwargs(test_fcn)
    argso = dili.get_default_kwargs(MyTestObj)
    argsi = dili.get_default_kwargs(myobj)
    logger.debug(f"argsf: {argsf}")
    assert argsf["c"] == 1
    assert argso["oc"] == 1
    assert argsi["oc"] == 1


class DictListTestCase(unittest.TestCase):
    def generate_dict_data(self):
        data = {"a": 1, "b": 2, "c": {"aa": 11, "bb": 22, "cc": {"aaa": 111}}}
        return data

    def test_ditc_flatten(self):
        data = {"a": 1, "b": 2, "c": {"aa": 11, "bb": 22, "cc": {"aaa": 111}}}
        dct = dili.flatten_dict(data)
        dct = dict(dct)
        self.assertIn("cccaaa", dct.keys())

    def test_ditc_flatten_keys_with_simplify(self):
        data = {"a": 1, "b": 2, "c": {"aa": 11, "bb": 22, "cc": {"aaa": [1, 2, 0]}}}
        dct = dili.flatten_dict_join_keys(data, simplify_iterables=True)
        dct = dict(dct)
        assert "c cc aaa 0" in dct.keys()
        assert dct["c cc aaa 0"] == 1

    def test_dict_flatten_with_separator(self):
        data = {"a": 1, "b": 2, "c": {"aa": 11, "bb": 22, "cc": {"aaa": 111}}}
        dct = dili.flatten_dict(data, separator=";")
        dct = dict(dct)
        self.assertIn("c;cc;aaa", dct.keys())

    def test_dict_recursive_update(self):
        data = self.generate_dict_data()
        data_updated = dili.recursive_update(data, {"c": {"aa": 33}})
        self.assertEqual(data_updated["c"]["aa"], 33)

    def test_list_to_dict(self):
        data = self.generate_dict_data()
        data["c"]["here is list"] = [1, 2, 5]

        data_with_dict = dili.list_to_dict_in_structure(data)
        dct = data_with_dict["c"]["here is list"]
        self.assertEqual(type(dct), dict)
        assert data_with_dict["c"]["here is list"][2] == 5

    def test_list_to_dict_key_str(self):
        data = self.generate_dict_data()
        data["c"]["here is list"] = [1, 2, 5]

        data_with_dict = dili.list_to_dict_in_structure(data, keys_to_str=True)
        dct = data_with_dict["c"]["here is list"]
        self.assertEqual(type(dct), dict)
        assert data_with_dict["c"]["here is list"]["2"] == 5

    def test_ndarray_to_list(self):
        data = self.generate_dict_data()
        data["c"]["here is ndarray"] = np.asarray([1, 2, 5])

        data_with_list = dili.ndarray_to_list_in_structure(data)
        self.assertEqual(type(data_with_list["c"]["here is ndarray"]), list)

    def test_dict_split(self):
        data = self.generate_dict_data()
        ab, c = dili.split_dict(data, ["a", "b"])
        self.assertIn("a", ab.keys())
        self.assertIn("b", ab.keys())
        self.assertIn("c", c.keys())

    def test_dict_subdict(self):
        data = self.generate_dict_data()

        ab = dili.subdict(data, ["a", "b"])
        self.assertIn("a", ab.keys())
        self.assertIn("b", ab.keys())
        # self.assertIn("c", c.keys())

    def test_split_dict_ordered(self):
        data = OrderedDict([("pear", 1), ("orange", 2), ("banana", 3), ("apple", 4)])
        ab, c = dili.split_dict(data, ["pear", "banana"])
        self.assertIn("pear", ab.keys())
        self.assertIn("banana", ab.keys())
        self.assertIn("apple", c.keys())

    def test_kick_from_dict(self):
        data = self.generate_dict_data()
        dct = dili.kick_from_dict(data, ["a", "b"])
        self.assertNotIn("a", dct.keys())
        self.assertNotIn("b", dct.keys())
        self.assertIn("c", dct.keys())

    def test_find_subsring_in_list(self):
        lst = ["auto", "veloco", "toto", "cola"]
        output = dili.list_contains(lst, "co")

        self.assertIn("veloco", output)
        self.assertIn("cola", output)

    def test_list_filter(self):
        lst = ["aa", "sss", "aaron", "rew"]
        output = dili.list_filter(lst, notstartswith="aa")
        self.assertTrue(["sss", "rew"] == output)

    def test_dict_find_key(self):
        slab = {"liver": 1, "porta": 2}
        val = dili.dict_find_key(slab, 2)
        self.assertEqual(val, "porta")

    # def test_dict_find_key_error(self):
    #     slab={"liver": 1, "porta": 2}
    #     val = dili.dict_find_key(slab, 3)
    #     self.assertEqual(val, "porta")
    def sort_list_data(self):
        dct = [
            {"name": "mira", "age": 34, "height": 172.0, "weight": 75},
            {"name": "kamca", "age": 25, "height": 152.0, "weight": 55},
            {"name": "bob", "age": 34, "height": 183.0, "weight": 85},
            {"name": "pavel", "age": 34, "height": 179.0, "weight": 98},
            {"name": "veru", "age": 25, "height": 162.0, "weight": 60},
            {"name": "pepa", "age": 39, "height": 182.0, "weight": 130},
        ]
        return dct

    def test_sort_list_of_dicts(self):
        dct = self.sort_list_data()

        dct = dili.sort_list_of_dicts(dct, keys=["age", "height"])
        self.assertEqual(dct[0]["name"], "kamca")
        self.assertEqual(dct[1]["name"], "veru")
        self.assertEqual(dct[2]["name"], "mira")
        self.assertEqual(dct[-1]["name"], "pepa")

    def test_sort_list_of_dicts_single_key(self):
        dct = self.sort_list_data()
        dct = dili.sort_list_of_dicts(dct, keys="height")
        self.assertEqual(dct[0]["name"], "kamca")
        # self.assertEqual(dct[1]["name"], "veru")
        # self.assertEqual(dct[2]["name"], "mira")
        self.assertEqual(dct[-1]["name"], "bob")

    def test_ordered_dict_to_dict(self):
        from collections import OrderedDict

        od = OrderedDict()
        od["klkj"] = 1
        od["here is dict"] = {"s": 1, 1: 17}
        od["list"] = ["uuu", 146, ["sdf", 18]]
        od2 = OrderedDict()
        od2["as"] = 1
        od2["as2"] = 1
        od["ordered dict"] = od2

        normal_dict = dili.ordered_dict_to_dict(od)
        self.assertEqual(type(normal_dict), dict)

    def test_drop_duplicates(self):
        import pandas as pd

        df = pd.DataFrame(
            {
                "col1": [5, 5, 1],
                "col2": [2, 2, 3],
                "time1": [1, 7, 9],
                "time2": [9, 7, 5],
            }
        )
        dfn = dili.df_drop_duplicates(df, "time")
        # self.assertEqual(len(dfn.keys()), 2)
        self.assertEqual(len(dfn), 2)

    def test_find_in_list_of_lists(self):
        lst = [[5, 6, 7], [1, 13], [2, 9]]
        self.assertEqual(dili.find_in_list_of_lists(lst, 7), 0)
        self.assertEqual(dili.find_in_list_of_lists(lst, 13), 1)
        self.assertEqual(dili.find_in_list_of_lists(lst, 9), 2)
        self.assertEqual(dili.find_in_list_of_lists(lst, 50), None)

    def test_find_value_in_struct(self):

        struct = {
            "bool": True,
            "int": 5,
            "str": "strdrr",
            "vs": [1.0, 2.5, 7],
            "data": {"complex": {"real": 1.0, "imag": 0.5}},
            "real": 1.1,
        }

        dfn = dili.find_value_in_struct(struct, 5)
        assert len(dfn) > 0
        # self.assertEqual(len(dfn.keys()), 2)
        # self.assertEqual(len(dfn), 2)

    def test_find_in_struct(self):
        struct = {
            "bool": True,
            "int": 5,
            "str": "strdrr",
            "vs": [1.0, 2.5, 7],
            "data": {"complex": {"real": 1.0, "imag": 0.5}},
            "real": 1.1,
        }

        dfn = dili.find_in_struct(struct, "bool")
        assert dfn == ["bool"]

        dfn = dili.find_in_struct(struct, ["complex", "data"])
        # this is probably not expected behavior
        assert dfn == ["data", "complex", "real"]

        dfn = dili.find_in_struct(struct, "real")
        assert dfn == ["data", "complex", "real"]

        val = dili.pick_from_struct(struct, dfn)
        val = 100

        dili.set_in_struct(struct, dfn, 10)
        assert struct["data"]["complex"]["real"] == 10


def test_find_closest_in_pandas():
    import pandas as pd

    df1 = pd.DataFrame({"x": [0.0, 0.0, 1.0, 1.0], "y": [0.0, 1.0, 1.0, 1.5],})
    df2 = pd.DataFrame({"x": [0.1, 1.1, 0.1, 1.1], "y": [0.1, 1.0, 1.0, 1.5],})

    ids = dili.find_id_of_nearest(df1, key=["x", "y"], df2=df2, key2=["x", "y"])
    assert all(list(ids) == np.asarray([0, 2, 1, 3]))

    ids = dili.find_id_of_nearest(df1, key=["x", "y"], df2=df2)
    assert all(list(ids) == np.asarray([0, 2, 1, 3]))

    # looking for closest in the same dataset
    ids = dili.find_id_of_nearest(df1, key=["x", "y"])
    assert all(list(ids) == np.asarray([1, 0, 3, 2]))


def main():
    unittest.main()


tests/exsu_test.py
#! /usr/bin/python
# -*- coding: utf-8 -*-

# import logging
# logger = logging.getLogger(__name__)
from loguru import logger
import pytest
import os.path

path_to_script = os.path.dirname(os.path.abspath(__file__))


def inc(x):
    return x + 1


# @pytest.mark.interactive
# @pytest.mark.slow
def test_answer():
    assert inc(3) == 4


tests/repo_test.py
#! /usr/bin/python
# -*- coding: utf-8 -*-

from loguru import logger
import unittest
import os
from pathlib import Path

# import shutil
import pandas as pd

# import openslide
import exsu.report
import exsu.git_tools


def test_repo():
    """
    Test non repo dir
    :return:
    """
    pth1 = Path(__file__).parent.parent.resolve().absolute()
    pth2 = Path(__file__).parent.parent.parent.resolve().absolute()
    logger.debug(f"pth1:{pth1}")
    logger.debug(f"pth2:{pth2}")
    rows_git_repo = exsu.git_tools.repo_status_to_dict(repodir=pth1)
    rows_not_git_repo = exsu.git_tools.repo_status_to_dict(repodir=pth2)
    assert len(rows_git_repo) > len(rows_not_git_repo)


def test_repo_in_report():
    report = exsu.report.Report(repodir=Path(__file__).parent.parent.resolve())
    report.finish_actual_row()
    print(report.df)
    assert report.df["repo exsu id"] is not None


exsu/git_tools.py
# /usr/bin/env python
# -*- coding: utf-8 -*-

from loguru import logger
from typing import Union
from pathlib import Path


def repo_status_to_dict(
    repodir: Union[Path, str], reponame: str = None, diff: bool = False
):
    """

    :param repodir:
    :param reponame:
    :param diff: Add diff string if true and if there are some changes.
    :return:
    """
    try:
        import git
        from git import Repo
    except ImportError as e:
        return {}
    repodir = Path(repodir)
    if is_git_repo(str(repodir)):
        repo = Repo(str(repodir))
    else:
        return {}

    if reponame is None:
        reponame = repodir.resolve().stem

    dirty = repo.is_dirty()
    retval = {
        f"repo {reponame} id": repo.head.object.hexsha,
        f"repo {reponame} is dirty": dirty,
    }
    if dirty:
        dirty_files = ",".join([item.a_path for item in repo.index.diff(None)])
        retval[f"repo {reponame} dirty files"] = dirty_files

    try:
        describe = repo.git.describe()
        retval[f"repo {reponame} describe"] = describe
    except git.exc.GitCommandError as e:
        logger.info(f"git describe error: {str(e)}")
    return retval


def is_git_repo(path):
    import git
    from git import Repo

    try:
        _ = git.Repo(path).git_dir
        return True
    except git.exc.InvalidGitRepositoryError:
        return False


exsu/report.py
# /usr/bin/env python
# -*- coding: utf-8 -*-
"""
Module provide simple interface to save experiment outputs directory with
spreadsheet and images.
"""
from loguru import logger
import pandas as pd
import os.path as op
import os
import warnings
from pathlib import Path
from typing import Union, List, Dict, Any
import numpy as np
from . import git_tools
from . import packages
import re


class Report:
    def __init__(
        self,
        outputdir=None,
        additional_spreadsheet_fn=None,
        level=50,
        save=True,
        show=True,
        debug=False,
        repodir=None,
        reponame=None,
        check_version_of: List[str] = None,
    ):
        """
        :param repodir: Directory with repository. The repository is checked when it is added.
        Repository hash will be added to every row.
        :param reponame: Name of repository. It can be derived automatically from repodir.
        :param outputdir: Directory for output. Have to be set for image report.
        :param additional_spreadsheet_fn:
        :param level: control initial level
        :param save:
        :param show:
        :param debug:
        :param check_version_of: list of strings with names of packages. The versions of the packages are saved to
        output spreadsheet when the line is finished.
        """
        # self.outputdir = op.expanduser(outputdir)

        self.df: pd.DataFrame = None
        # self.imgs: Dict[str, np.ndarray] = {}
        self.imgs: dict = {}
        self.actual_row: dict = {}
        self.show = show
        self.save = save
        self.debug = debug
        self.level = level
        self.spreadsheet_fn = "data.xlsx"
        self.additional_spreadsheet_fn = additional_spreadsheet_fn
        self.persistent_cols: dict = {}
        # self.repos = []
        self.outputdir = None
        self.check_version_of = check_version_of

        if outputdir is not None:
            self.init_with_output_dir(outputdir)
        else:
            self.init()
        self.repos: dict = {}

        if repodir is not None:
            self.add_repo(repodir, reponame)

    def add_repo(self, repodir, reponame: str = None):
        # self.repos.push((repodir, reponame))
        self.repos[reponame] = Path(repodir)
        repo_cols = git_tools.repo_status_to_dict(repodir, reponame)
        self.set_persistent_cols(repo_cols)

    def set_persistent_cols(self, dct: dict, clear: bool = False):
        """
        Set data which will be appended to all rows.
        :param dct: dictionary with column name and value
        :param clear: True or False. Clear persistent cols before setting.
        :return:
        """
        if clear:
            self.persistent_cols = dct
        else:
            self.persistent_cols.update(dct)

        logger.debug(f"Adding persistent cols: {list(self.persistent_cols.keys())}")

    def init_with_output_dir(self, outputdir):
        """
        Make set_output_dir() and init report.
        :param outputdir:
        :return:
        """
        self.set_output_dir(outputdir)
        self.init()

    def set_output_dir(self, outputdir):
        """
        Set the output dir with expanrequired. .
        :param outputdir:
        :return:
        """
        self.outputdir = Path(outputdir).expanduser()
        # if not op.exists(self.outputdir):
        #     os.makedirs(self.outputdir)

    def set_show(self, show):
        self.show = show

    def set_save(self, save):
        self.save = save

    def add_cols_to_actual_row(self, data):
        self.actual_row.update(data)

    # def write_table(self, filename):
    def finish_actual_row(self):
        data = self.actual_row
        logger.trace(f"Actual row cols: {list(self.actual_row.keys())}")
        logger.trace(f"Persistent cols: {list(self.persistent_cols.keys())}")
        data.update(self.persistent_cols)
        if self.check_version_of is not None:
            data.update(packages.get_version_of_packages_as_dict(self.check_version_of))

        df = pd.DataFrame([list(data.values())], columns=list(data.keys()))
        logger.trace(
            f"Unique values types {np.unique(map(str, map(type, data.values())))}"
        )
        self.df = self.df.append(df, ignore_index=True)

        # if excel_path.exists():
        #     print("append to excel")
        #
        # else:
        #
        #     # writer = pd.ExcelWriter(filename, engine='openpyxl')
        #     df.to_excel(filename)
        #     print("create new excel")

        self.actual_row = {}

    # def add_table(self):
    #     pass

    def init(self):
        self.df = pd.DataFrame()
        self.imgs = {}
        self.actual_row = {}

    def dump(self):
        if self.outputdir is not None:
            pth = self.join_output_dir(self.spreadsheet_fn)
            logger.debug(f"Saving to file {pth}")
            ppth = Path(pth)
            if not ppth.parent.exists():
                logger.debug("creating output dir")
                os.makedirs(ppth.parent)
            self.df.to_excel(pth, index=False)

        if self.additional_spreadsheet_fn is not None:
            excel_path = Path(self.additional_spreadsheet_fn)
            # print("we will write to excel", excel_path)
            filename = str(excel_path)
            append_df_to_excel(filename, self.df)
        logger.debug(f"Saved")
        # append_df_to_excel_no_head_processing(filename, self.df)

    def imsave(
        self,
        base_fn,
        arr: np.ndarray,
        level=90,
        level_skimage=40,
        level_npz=20,
        k=50,
        kwargs_skimage=None,
        **kwargs,
    ):
        """
        Save image to report dir.

        * The usual way is to use matplotlib to produce colored images from grayscale with no\
        additional stuff (like axes).

        * The alternative is to skimage to produce real grayscale. The\
        intensity values can easily overflow the uint8 range. It can be controled by the
        parameter 'k'.

        * The RAW data can be stored into .npz file format.

        :param base_fn: with a format slot for annotation id like "skeleton_{}.png"
        :param arr: numpy array with input data
        :param level: severity of input
        :param level_skimage: severity of input. It is not colloring the image.
        :param k: multiplier applied on input image intensities on skimage save
        :param kwargs_skimage: used for internal call of skimage.io.imsave
        :param kwargs are used for internal call plt.imsave
        :return:
        """

        import skimage.io

        if kwargs_skimage is None:
            kwargs_skimage = {}

        filename, ext = op.splitext(base_fn)
        if ext == "":
            ext = ".png"
        if self.save:
            if self._is_under_level(level):
                import matplotlib.pyplot as plt

                fn = self.join_output_dir(filename + ext)
                logger.debug(f"write to file: {fn}")
                plt.imsave(fn, arr, **kwargs)

            if self._is_under_level(level_skimage):
                with warnings.catch_warnings():
                    warnings.filterwarnings("ignore", ".*low contrast image.*")
                    ## warnings.simplefilter("low contrast image")
                    fn = self.join_output_dir(filename + "_skimage" + ext)
                    # if "check_contrast" not in kwargs_skimage:
                    #     kwargs_skimage["check_contrast"] = False
                    logger.debug(f"write to file: {fn}")
                    skimage.io.imsave(fn, k * arr, *kwargs_skimage)
            if self._is_under_level(level_npz):
                self._save_arr(base_fn, arr)

    def join_output_dir(self, *args, **kwargs):
        """
        Join path to output dir and create output dir if necessary
        :param args:
        :param kwargs:
        :return:
        """
        if self.outputdir is None:
            raise ValueError("Outputdir have to be set before file is saved.")
        if not op.exists(self.outputdir):
            os.makedirs(self.outputdir)
        return op.join(self.outputdir, *args, **kwargs)

    def _save_arr(self, base_fn, arr: np.ndarray):
        """
        Save ndarray to file and store file path to list
        :param base_fn:
        :param arr:
        :return:
        """
        filename, ext = op.splitext(base_fn)
        fn = self.join_output_dir(filename + ".npz")
        np.savez_compressed(fn, arr=arr)
        self.imgs[base_fn] = fn

    def load_array(self, base_fn) -> np.ndarray:
        """

        :param base_fn: read image source array from file based on base file name
        :return:
        """
        arr = np.load(self.imgs[base_fn])["arr"]
        # logger.debug(arr)

        return arr

    def imsave_as_fig(self, base_fn, arr, level=60, npz_level=30, **kwargs_savefig):
        """
        Save given array as figure and save array as npy as well.
        :param base_fn:
        :param arr:
        :param level:
        :return:
        """
        filename, ext = op.splitext(base_fn)
        if self._is_under_level(level):
            import matplotlib.pyplot as plt

            fig = plt.figure()
            plt.imshow(arr)
            plt.colorbar()
            if self.save:
                fn = self.join_output_dir(filename + "" + ext)
                plt.savefig(fn, **kwargs_savefig)
            if self.show:
                plt.show()
            else:
                plt.close(fig)
        if self._is_under_level(npz_level):
            self._save_arr(base_fn, arr)

    # def add_array(self, base_fn, arr, k=50):
    #     if self.save:
    #         self.imsave

    def savefig(self, base_fn, level=60, **kwargs_savefig):
        """
        Save figure with matploglib if severity is high enough.
        Do not close figure.

        :param base_fn: filename
        :param level: severity of input data
        :return:
        """
        import matplotlib.pyplot as plt

        if self._is_under_level(level):
            filename, ext = op.splitext(base_fn)
            if ext == "":
                ext = ".png"
            if self.save:
                fn = self.join_output_dir(filename + "" + ext)
                # plt.show()
                logger.trace(kwargs_savefig)
                logger.debug(fn)
                plt.savefig(fn, **kwargs_savefig)
                # self.imgs[base_fn] = [fn]

    def savefig_no_border(self, base_fn, level=60, **kwargs_savefig):
        import matplotlib.pyplot as plt
        plt.gca().set_axis_off()
        plt.subplots_adjust(top=1, bottom=0, right=1, left=0,
                            hspace=0, wspace=0)
        plt.margins(0, 0)
        plt.gca().xaxis.set_major_locator(plt.NullLocator())
        plt.gca().yaxis.set_major_locator(plt.NullLocator())
        self.savefig(base_fn=base_fn, level=level, bbox_inches='tight',
                    pad_inches=0, **kwargs_savefig)


    def savefig_and_show(self, base_fn, fig, level=60, **kwargs_savefig):
        """
        Save figure with matploglib if severity is high enough and then show if
        report 'show' parameter is set true.

        :param base_fn: filename
        :param fig: figure id (used for closing figure if show is false)
        :param level: severity of input data
        :return:
        """
        import matplotlib.pyplot as plt

        self.savefig(base_fn, level=level, **kwargs_savefig)
        if self.show:
            plt.show()
        else:
            plt.close(fig)

    # def save_np_data(self, base_fn, data, format_args=None, level=60):
    #     if self._is_under_level(level):
    #         if format_args is None:
    #             format_args = []
    #         fn = self._imjoin(self.outputdir, base_fn.format(format_args))
    #         np.save(data, fn)
    #         self.imgs[base_fn] = fn

    def _is_under_level(self, level):
        return level_numeric_value(self.level) < level_numeric_value(level)


def level_numeric_value(level):
    """
    Convert string levels like 'debug' to number.
    :param level:
    :return:
    """
    if level is "debug":
        level = 10
    elif level is "info":
        level = 30
    elif level is "warning":
        level = 50
    return level


def append_df_to_excel(
    filename,
    df,
    sheet_name="Sheet1",
    startrow=None,
    truncate_sheet=False,
    **to_excel_kwargs,
):
    """
    Append a DataFrame [df] to existing Excel file [filename]
    into [sheet_name] Sheet.
    If [filename] doesn't exist, then this function will create it.

    Excel file have to be closed.

    Parameters:
      filename : File path or existing ExcelWriter
                 (Example: '/path/to/file.xlsx')
      df : dataframe to save to workbook
      sheet_name : Name of sheet which will contain DataFrame.
                   (default: 'Sheet1')
      startrow : upper left cell row to dump data frame.
                 Per default (startrow=None) calculate the last row
                 in the existing DF and write to the next row...
      truncate_sheet : truncate (remove and recreate) [sheet_name]
                       before writing DataFrame to Excel file
      to_excel_kwargs : arguments which will be passed to `DataFrame.to_excel()`
                        [can be dictionary]

    Returns: pandas dataframe with combined information
    """
    logger.debug(f"Writing XLSX to: {str(filename)}")
    # from openpyxl import load_workbook

    import pandas as pd

    filename = Path(filename)
    dfo = df
    if filename.exists():
        dfold = read_spreadsheet(filename, sheet_name=sheet_name)
        # dfout = pd.concat([dfin, df], axis=0, ignore_index=True)
        dfo = dfold.append(df, ignore_index=True, sort=True)
        # df.to_excel(str(filename), sheet_name=sheet_name, index=False, engine='openpyxl')
        # try:
        #     dfold = pd.read_excel(str(filename), sheet_name=sheet_name)
        #     dfcombine = dfold.append(df, ignore_index=True)
        #     dfcombine.to_excel(str(filename), sheet_name=sheet_name)
        # except PermissionError as e:
        # print("File is opened in other application")
        # import xlwings as xw
        # sht = xw.Book(str(filename)).sheets[0]
        # sht.range('A1').expand().options(pd.DataFrame).value
    try:
        dfo = _write_spreadsheet(dfo, filename, sheet_name)
    except PermissionError as e:
        srch = re.search(r".*_([0-9]+)", filename.stem)
        logger.debug(srch)
        if srch is None:
            number = 0
        else:
            number = int(srch.groups()[0])

        number += 1

        fn2 = filename.parent / (filename.stem + f".{number}" + filename.suffix )
        logger.warning(f"Permission Error. File {str(filename)} already used. Trying save to {str(fn2)}")
        dfo = append_df_to_excel(fn2, df=df, sheet_name=sheet_name)
    return dfo



def _write_spreadsheet(df:pd.DataFrame, filename:Path, sheet_name:str):
    filename.parent.mkdir(parents=True, exist_ok=True)
    # pd.read_excel(filename, sheet_name=)
    if filename.suffix == ".xlsx":
        df.to_excel(
            str(filename), sheet_name=sheet_name, index=False, engine="openpyxl"
        )
    elif filename.suffix == ".xls":
        df.to_excel(str(filename), sheet_name=sheet_name, index=False)
    elif filename.suffix == ".csv":
        df.to_csv(str(filename), index=False)
    else:
        raise ValueError(f"Suffix '{filename.suffix}' is not supported.")
    return df

    # # ignore [engine] parameter if it was passed
    # if 'engine' in to_excel_kwargs:
    #     to_excel_kwargs.pop('engine')
    #
    # writer = pd.ExcelWriter(filename, engine='openpyxl')
    #
    # # Python 2.x: define [FileNotFoundError] exception if it doesn't exist
    # try:
    #     FileNotFoundError
    # except NameError:
    #     FileNotFoundError = IOError
    #
    #
    # try:
    #     # try to open an existing workbook
    #     writer.book = load_workbook(filename)
    #
    #     # get the last row in the existing Excel sheet
    #     # if it was not specified explicitly
    #     if startrow is None and sheet_name in writer.book.sheetnames:
    #         startrow = writer.book[sheet_name].max_row
    #
    #     # truncate sheet
    #     if truncate_sheet and sheet_name in writer.book.sheetnames:
    #         # index of [sheet_name] sheet
    #         idx = writer.book.sheetnames.index(sheet_name)
    #         # remove [sheet_name]
    #         writer.book.remove(writer.book.worksheets[idx])
    #         # create an empty sheet [sheet_name] using old index
    #         writer.book.create_sheet(sheet_name, idx)
    #
    #     # copy existing sheets
    #     writer.sheets = {ws.title:ws for ws in writer.book.worksheets}
    # except FileNotFoundError:
    #     # file does not exist yet, we will create it
    #     pass
    #
    # if startrow is None:
    #     startrow = 0
    #
    # # write out the new sheet
    # df.to_excel(writer, sheet_name, startrow=startrow, **to_excel_kwargs)
    #
    # # save the workbook
    # writer.save()


def read_spreadsheet(filename: Union[str, Path], sheet_name="Sheet1"):
    filename = Path(filename)
    if filename.suffix == ".xls":
        # writer = pd.ExcelWriter(filename, engine='openpyxl')
        dfold = pd.read_excel(str(filename), sheet_name=sheet_name)
    elif filename.suffix == ".xlsx":
        dfold = pd.read_excel(str(filename), sheet_name=sheet_name, engine="openpyxl")
    elif filename.suffix == ".csv":
        dfold = pd.read_csv(str(filename))
    else:
        raise ValueError(f"Suffix '{filename.suffix}' is not supported.")
    return dfold


# def append_df_to_excel_no_head_processing(filename, df, sheet_name='Sheet1', startrow=None,
#                                           truncate_sheet=False,
#                                           **to_excel_kwargs):
#     """
#
#     Append a DataFrame [df] to existing Excel file [filename]
#     into [sheet_name] Sheet.
#     If [filename] doesn't exist, then this function will create it.
#     It does insert also first line with column name :-(
#
#     Parameters:
#       filename : File path or existing ExcelWriter
#                  (Example: '/path/to/file.xlsx')
#       df : dataframe to save to workbook
#       sheet_name : Name of sheet which will contain DataFrame.
#                    (default: 'Sheet1')
#       startrow : upper left cell row to dump data frame.
#                  Per default (startrow=None) calculate the last row
#                  in the existing DF and write to the next row...
#       truncate_sheet : truncate (remove and recreate) [sheet_name]
#                        before writing DataFrame to Excel file
#       to_excel_kwargs : arguments which will be passed to `DataFrame.to_excel()`
#                         [can be dictionary]
#
#     Returns: None
#     """
#     # from openpyxl import load_workbook
#
#     import pandas as pd
#
#     # ignore [engine] parameter if it was passed
#     if 'engine' in to_excel_kwargs:
#         to_excel_kwargs.pop('engine')
#
#     writer = pd.ExcelWriter(filename, engine='openpyxl')
#
#     # Python 2.x: define [FileNotFoundError] exception if it doesn't exist
#     try:
#         FileNotFoundError
#     except NameError:
#         FileNotFoundError = IOError
#
#
#     try:
#         # try to open an existing workbook
#         from openpyxl import load_workbook
#         writer.book = load_workbook(filename)
#
#         # get the last row in the existing Excel sheet
#         # if it was not specified explicitly
#         if startrow is None and sheet_name in writer.book.sheetnames:
#             startrow = writer.book[sheet_name].max_row
#
#         # truncate sheet
#         if truncate_sheet and sheet_name in writer.book.sheetnames:
#             # index of [sheet_name] sheet
#             idx = writer.book.sheetnames.index(sheet_name)
#             # remove [sheet_name]
#             writer.book.remove(writer.book.worksheets[idx])
#             # create an empty sheet [sheet_name] using old index
#             writer.book.create_sheet(sheet_name, idx)
#
#         # copy existing sheets
#         writer.sheets = {ws.title:ws for ws in writer.book.worksheets}
#     except FileNotFoundError:
#         # file does not exist yet, we will create it
#         pass
#
#     if startrow is None:
#         startrow = 0
#
#     # write out the new sheet
#     df.to_excel(writer, sheet_name, startrow=startrow, **to_excel_kwargs)
#
#     # save the workbook
#     writer.save()
